packages/astro/src/core/build/plugins/plugin-content.ts

```diff
@@ -0,0 +1,302 @@
+							const resultId = encodeName(`${removeLeadingForwardSlash(removeFileExtension(srcRelativePath))}.render.mjs`);
+							return resultId;
+						}
+						const [srcRelativePath, flag] = id.replace(rootPath, '/').split('?');
+						const collectionEntry = findEntryFromSrcRelativePath(lookupMap, srcRelativePath);
+						if (collectionEntry) {
+							let suffix = '.mjs';
+							if (flag === PROPAGATED_ASSET_FLAG) {
+								suffix = '.entry.mjs';
+							}
+							id = removeLeadingForwardSlash(removeFileExtension(encodeName(id.replace(srcPath, '/')))) + suffix;
+							return id;
+						}
+					}
+				}
+			});
+		},
+
+		resolveId(id) {
+			if (id === virtualEmptyModuleId) {
+				return resolvedVirtualEmptyModuleId;
+			}
+		},
+
+		async load(id) {
+			if (id === resolvedVirtualEmptyModuleId) {
+				return {
+					code: `// intentionally left empty!\nexport default {}`
+				}
+			}
+		},
+
+		async generateBundle(_options, bundle) {
+			const code = await generateContentEntryFile({ settings: opts.settings, fs: fsMod, lookupMap, IS_DEV: false, IS_SERVER: false });
+			this.emitFile({
+				type: 'prebuilt-chunk',
+				code,
+				fileName: 'content/entry.mjs'
+			})
+			if (!injectedEmptyFile) return;
+			Object.keys(bundle).forEach(key => {
+				const mod = bundle[key];
+				if (mod.type === 'asset') return;
+				if (mod.facadeModuleId === resolvedVirtualEmptyModuleId) {
+					delete bundle[key];
+				}
+			});
+		},
+
+		async writeBundle() {
+			// These are stored in the manifest to ensure that they are included in the build
+			// in case they aren't referenced _outside_ of the cached content.
+			// We can use this info in the manifest to run a proper client build again.
+			const clientComponents = new Set([
+				...oldManifest.clientEntries,
+				...internals.discoveredHydratedComponents.keys(),
+				...internals.discoveredClientOnlyComponents.keys(),
+				...internals.discoveredScripts,
+			])
+			// Likewise, these are server modules that might not be referenced
+			// once the cached items are excluded from the build process
+			const serverComponents = new Set([
+				...oldManifest.serverEntries,
+				...internals.discoveredHydratedComponents.keys(),
+			]);
+			newManifest.serverEntries = Array.from(serverComponents);
+			newManifest.clientEntries = Array.from(clientComponents);
+			await fsMod.promises.mkdir(contentCacheDir, { recursive: true });
+			await fsMod.promises.writeFile(contentManifestFile, JSON.stringify(newManifest), { encoding: 'utf8' });
+
+			const cacheExists = fsMod.existsSync(cache);
+			fsMod.mkdirSync(cache, { recursive: true })
+			await fsMod.promises.mkdir(cacheTmp, { recursive: true });
+			await copyFiles(distContentRoot, cacheTmp, true);
+			if (cacheExists) {
+				await copyFiles(contentCacheDir, distContentRoot, false);
+			}
+			await copyFiles(cacheTmp, contentCacheDir);
+			await fsMod.promises.rm(cacheTmp, { recursive: true, force: true });
```

Comment:
Can we enclose all these FS operations in a `try/catch` block? There might be remote situations where we don't have write access